src/Cardano/Wallet.hs

```diff
@@ -72,12 +72,23 @@ data WalletLayer s = WalletLayer
     { createWallet
         :: NewWallet
         -> ExceptT CreateWalletError IO WalletId
+        -- ^ Initialise and store a new wallet, returning its ID.
     , readWallet
         :: WalletId
         -> ExceptT ReadWalletError IO (Wallet s, WalletMetadata)
+        -- ^ Retrieve the wallet state for the wallet with the given ID.
     , watchWallet
-        :: WalletId
+        :: forall a. WalletId
+        -> IO a
+        -- ^ Consume blocks from the node as they arrive, and apply them to the
+        -- wallet. This function never returns, but may throw an exception.
+    , processWallet
+        :: (SlotId -> IO ())
+        -> WalletId
         -> IO ()
+        -- ^ Consume the entire available chain, applying block to the given
+        -- wallet, and stop when finished. This function is intended for
+        -- benchmarking.
```

Comment:
This is very slippery slope that has caused us a lot of pain in the previous implementations. Let's **not** introduce code that is test-specific in our codebase. If something is hard to test, I'd rather review our source implementation than introduce tests work-around..
In the end, we do want to test our production code, and not some testing variation of it.

Here's what I'd suggest:

- Get rid of `watchWallet` & `processWallet` altogether. 
- Introduce a new `restoreWallet :: WalletId -> (SlotId, SlotId) -> ExceptT RestoreWalletError IO ()` which would restore a wallet from a starting slot up to another slot.
- We can pretty much re-use the existing `applyBlocks` to do so, and simply, add an extra round of getting the "next" slot from the network layer.

Reply:
Thanks for the lecture, I agree with not introducing test specific code, and that implementing tests improves the API - however -
1. The problem with the previous implementation was a complete absence of benchmarks, not test-specific code.
2. There are no call sites of `watchWallet` yet and at the moment there is no clear spec of how this function or `restoreWallet` will be used. I suggest fixing it not now but soon when the caller of `restoreWallet` is written.
3. In the interests of small short-lived PRs, this change should be done in a separate PR.

Reply:
1. The problem was, unfortunately, also test-specific code. We had entire modules that were almost only present for testing. And functions with arbitrarily complex signatures to allow passing around custom data and short-circuit surrounding logic :( ... 

2. & 3. True, hence why I think we could remove it. We probably added it too early for "testing" whereas in the end, there was still much work to do before that. However, your suggestion is sensible. I'll get it merge as it is and issue a follow-up PR :+1: